src/core/lib/security/transport/security_connector.cc

```diff
@@ -704,18 +704,31 @@ static void ssl_server_add_handshakers(grpc_server_security_connector* sc,
 }
 
 static int ssl_host_matches_name(const tsi_peer* peer, const char* peer_name) {
-  char* allocated_name = nullptr;
+  char* allocated_host = nullptr;
+  char* allocated_addr = nullptr;
   int r;
 
   if (strchr(peer_name, ':') != nullptr) {
     char* ignored_port;
-    gpr_split_host_port(peer_name, &allocated_name, &ignored_port);
+    gpr_split_host_port(peer_name, &allocated_host, &ignored_port);
     gpr_free(ignored_port);
-    peer_name = allocated_name;
+    peer_name = allocated_host;
     if (!peer_name) return 0;
   }
+
+  /* IPv6 non-global zone-id should not be included in comparisons. */
+  const char* percent = strchr(peer_name, '%');
```

Comment:
Could `%` possibly mean anything else in this context?  I'm thinking about cases like `google.com%2Eattacker.evil` or `evil.news%2Egoogle.com`.

Every IPv6 address contains a `:`, so you could avoid a malloc by truncating `allocated_host`.

In general, an IPv6 link-local address is almost as low-level as a MAC address: every address in the world shares the same /64.  Is it actually plausible that anyone would operate TLS at this layer in the stack?

Reply:
I work on a networking product. We'd like to be able to securely diagnose issues with network configuration when higher levels have potentially failed. So that's one use-case. I agree I can't think of many others off the top of my head. 

re `%` in domain names traditionally they're just alphanumeric and hyphen. But there's been some more recent push regarding support for unicode in hostnames https://en.wikipedia.org/wiki/Punycode and I don't know if gRPC a) supports this and b) would perform the conversion to plain ascii when checking alternate names.

Reply:
@pmarks-net re allocation: I didn't want to modify `allocated_name` from `grpc_split_host_port` but now that you mention it I can't think of anything bad about that approach and the fewer allocations the better. Thanks!

Reply:
> We'd like to be able to securely diagnose issues with network configuration when higher levels (slaac, dhcp, dns, etc.) have potentially failed.

Will this change actually allow you to do that?  If I understand correctly, you would need to obtain TLS certificates for your link-local addresses, which sounds very strange to me. Does gRPC work well with self-signed certificates?

Reply:
> Does gRPC work well with self-signed certificates?

Yep! You can specify your own list of root trust anchors. We are self-issuing certificates during our provisioning stage and compiling the trust roots for use in our application layer (which uses gRPC). I have verified that with this fix I'm able to connect to a service on one of my hosts via its link-local address without explicitly overriding the expected host name.

Reply:
rgr I will update/simplify when I fix the conflicts. Stay tuned.